src/Google.Api.Gax.Grpc/CallSettings.cs

```diff
@@ -116,22 +119,26 @@ internal static CallSettings Merge(CallSettings original, CallSettings overlaid)
         internal static CallOptions ToCallOptions(CallSettings baseSettings, CallSettings callSettings, IClock clock)
         {
             CallSettings effectiveSettings = CallSettings.Merge(baseSettings, callSettings);
-            return effectiveSettings == null
-                ? default(CallOptions)
-                : new CallOptions(
-                    headers: effectiveSettings.Headers,
-                    deadline: effectiveSettings.Timing.CalculateDeadline(clock),
-                    cancellationToken: effectiveSettings.CancellationToken ?? default(CancellationToken),
-                    writeOptions: effectiveSettings.WriteOptions,
-                    propagationToken: effectiveSettings.PropagationToken,
-                    credentials: effectiveSettings.Credentials);
+            if (effectiveSettings == null)
+            {
+                return default(CallOptions);
+            }
+            var metadata = new Metadata();
+            effectiveSettings.HeaderMutation?.Invoke(metadata);
+            return new CallOptions(
+                headers: metadata,
+                deadline: effectiveSettings.Timing.CalculateDeadline(clock),
+                cancellationToken: effectiveSettings.CancellationToken ?? default(CancellationToken),
+                writeOptions: effectiveSettings.WriteOptions,
+                propagationToken: effectiveSettings.PropagationToken,
+                credentials: effectiveSettings.Credentials);
         }
 
         /// <summary>
         /// Creates a <see cref="CallSettings"/> for the specified user agent.
         /// </summary>
         internal static CallSettings ForUserAgent(string userAgent) =>
-            new CallSettings(null, null, null, new Metadata { { UserAgentBuilder.HeaderName, userAgent } }, null, null);
+            new CallSettings(null, null, null, metadata => metadata.Add(UserAgentBuilder.HeaderName, userAgent), null, null);
```

Comment:
Add a CallSettings.FromMetaDataMutator(...) ?
This would be consistent with all the other From...() methods.